<div style="text-align:center;">
  <span style="color:Green; font-size:30px; font-weight:bold;">
    Bad Debt Prediction
  </span>
</div>

---------------------------------------------

**I cleaned and prepared the data by merging datasets, correcting improper text values, and converting raw data into meaningful business terms.**

--------------------------------------------------

In [150]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
import warnings

warnings.filterwarnings('ignore')

warnings.filterwarnings('default')



pd.set_option('display.float_format', lambda x: '{:.2f}'.format(x))
np.set_printoptions(suppress=True)

## **Loading The Data**

## **Data Table**

In [154]:
df_data = pd.read_excel("Data.xlsm")
df_data.head()

,OUTCOME,Customer,RISK_SCORE,BUREAU_ENQUIRIES_12_MONTHS,LICENCE_AVAILABLE,RESIDENTIAL,BUREAU_DEFAULT,CD_OCCUPATION,CD_STS_RESIDENTIAL,DEFAULT_AMT,ENQUIRIES_6M,ENQUIRIES_12M,DEROGATORIES,DEFAULTS_12M,NO_MONTHS_VCR221,NO_STATUS_VCR183,CD_STS_VCR184,NO_MONTHS_VCR228,NO_MONTHS_VCR226,NO_SUM_VCR053,NO_MONTHS_VCR258,SCORECARD,NO_SCORE
0,Good,2975536526,628,queries 12 Months 1-2,Driver Licence Check Missing,Rented,Default Current- Missing,E,R,0,2,2,0,0,0,-,-,0,0,0,0,TAR1A,724
1,Good,2999806086,813,queries 12 Months 1-2,Driver Licence Check Yes,Rented,Default Current- Missing,E,R,0,2,2,0,0,0,1,-,24,0,4000,23,TAR1A,775
2,Good,2800835182,1012,queries 12 Months 4-5,Driver Licence Check Yes,Owned,Default Current- Missing,E,O,0,5,5,0,0,0,0,-,24,0,380277,24,SFJR1A,815
3,Good,2975640662,415,queries 12 Months 4-5,Driver Licence Check Missing,Rented,Default Current1-1000,E,R,459,2,4,1,0,0,-,-,0,0,0,0,HSHSOL,328
4,Good,2979232086,556,queries 12 Months 1-2,Driver Licence Check Missing,Board or Living With Relative,Default Current- Missing,U,B,0,2,2,0,0,0,-,-,0,0,0,0,TAR1A,692


In [155]:
df_data.columns

Index(['OUTCOME', 'Customer', 'RISK_SCORE', 'BUREAU_ENQUIRIES_12_MONTHS',
       'LICENCE_AVAILABLE', 'RESIDENTIAL', 'BUREAU_DEFAULT', 'CD_OCCUPATION',
       'CD_STS_RESIDENTIAL', 'DEFAULT_AMT', 'ENQUIRIES_6M', 'ENQUIRIES_12M',
       'DEROGATORIES', 'DEFAULTS_12M', 'NO_MONTHS_VCR221', 'NO_STATUS_VCR183',
       'CD_STS_VCR184', 'NO_MONTHS_VCR228', 'NO_MONTHS_VCR226',
       'NO_SUM_VCR053', 'NO_MONTHS_VCR258', 'SCORECARD', 'NO_SCORE'],
      dtype='object')

In [156]:
df_data.isnull().sum()

OUTCOME                       0
Customer                      0
RISK_SCORE                    0
BUREAU_ENQUIRIES_12_MONTHS    0
LICENCE_AVAILABLE             0
RESIDENTIAL                   0
BUREAU_DEFAULT                0
CD_OCCUPATION                 0
CD_STS_RESIDENTIAL            0
DEFAULT_AMT                   0
ENQUIRIES_6M                  0
ENQUIRIES_12M                 0
DEROGATORIES                  0
DEFAULTS_12M                  0
NO_MONTHS_VCR221              0
NO_STATUS_VCR183              0
CD_STS_VCR184                 0
NO_MONTHS_VCR228              0
NO_MONTHS_VCR226              0
NO_SUM_VCR053                 0
NO_MONTHS_VCR258              0
SCORECARD                     0
NO_SCORE                      0
dtype: int64

In [157]:
df_data['Customer'].duplicated().sum()

np.int64(0)

## **Additional_Information**

In [158]:
df_ad = pd.read_excel("Additional_Information.xlsx")
df_ad.head()

,Customer,SCORE_INDICATOR_CR22,SCORE_CR22,SCORE_BAND_CR22,SCORECARD_CR21,NO_SCORE,RISK_RELATIVE_CR21,RISK_RELATIVE_ODDS_CR21,NO_MONTHS_OLDEST_ACCOUNT,OPEN_ACCT_CR22,CLOSED_ACCT_CR22,MORTGAGE_CR22,CREDIT_CARD_CR22,TELCO_CR22,UTILITIES_CR22,OVERDRAFT_CR22,PERSONAL_LOAN_CR22,AUTO_LOAN_CR22,CREDIT_CARD_AMT_CR22,OVERDRAFT_AMT_CR22,PERSONAL_LOAN_AMT_CR22,AUTO_LOAN_AMT_CR22,OPEN_ACCT_CR21,CLOSED_ACCT_CR21,MORTGAGE_CR21,CREDIT_CARD_CR21,TELCO_CR21,UTILITIES_CR21,OVERDRAFT_CR21,PERSONAL_LOAN_CR21,AUTO_LOAN_CR21,CREDIT_CARD_AMT_CR21,OVERDRAFT_AMT_CR21,PERSONAL_LOAN_AMT_CR21,AUTO_LOAN_AMT_CR21,ENQUIRY_CNT_CR22,ENQUIRY_CNT_1M_CR22,ENQUIRY_CNT_3M_CR22,ENQUIRY_CNT_6M_CR22,ENQUIRY_CNT_12M_CR22,ENQUIRY_CNT_CR21,ENQUIRY_CNT_1M_CR21,ENQUIRY_CNT_3M_CR21,ENQUIRY_CNT_6M_CR21,ENQUIRY_CNT_12M_CR21,ENQUIRY_MAX_AMT_1M,ENQUIRY_MAX_AMT_3M,ENQUIRY_MAX_AMT_6M,ENQUIRY_MAX_AMT_12M,DEFAULT_CNT_CR22,DEFAULT_OPEN_CNT_CR22,DEFAULT_CNT_3M_CR22,DEFAULT_CNT_6M_CR22,DEFAULT_CNT_12M_CR22,DEFAULT_UNPAID_AMT_CR22,DEFAULT_CNT_CR21,DEFAULT_OPEN_CNT_CR21,DEFAULT_UNPAID_AMT_CR21,DEFAULT_CNT_3M_CR21,DEFAULT_CNT_6M_CR21,DEFAULT_CNT_12M_CR21,WORST_1M,WORST_3M,WORST_6M,WORST_12M,WORST_18M,WORST_24M,HAD_SERVICES,DOC_TYPE,VISA_TYPE,EMPLOYED_STATUS,EMPLOYED_TYPE,EMPLOYED_NO_MTHS,NO_ADDRESS_AT_MONTHS,CD_STS_RESIDENTIAL,AGENCY_AMT,MTH_TILL_AGENCY
0,2984743758,Clean Positive,373.00,H,CTS1A,608.00,4.09,16.91,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,Intl Passport and Visa,Work Visa,E,F,3.00,0.00,R,NaN,NaN
1,2984205798,Clean Positive,1024.00,A,CTSSH,827.00,18.66,77.17,268,2,0,0,2,0,0,0,0,0,18000,0,0,0,3,0,0,0,1,0,0,0,0,0,0,0,0,2,1,1,1,1,2,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,AU Driver Licence,Resident,R,U,0.00,120.00,O,NaN,NaN
2,2983553714,Clean Positive,1005.00,A,CTSSH,833.00,19.45,80.45,339,6,1,5,1,0,0,0,0,0,3000,0,0,0,6,1,5,0,0,0,0,0,0,0,0,0,0,2,1,1,1,1,3,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,AU Driver Licence,Resident,E,F,131.00,24.00,O,NaN,NaN
3,2972077238,Clean Positive,506.00,G,CTS1A,724.00,9.14,37.79,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,AU Driver Licence,Resident,E,P,2.00,2.00,R,NaN,NaN
4,2975514910,Dirty Positive,98.00,H,CTSDP,484.00,1.73,7.16,4,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6,1,1,2,5,7,1,1,4,5,0,0,529,4500,1,1,0,0,0,726,0,0,0,0,0,0,3,3,3,3,3,3,1,AU Driver Licence,Resident,E,F,20.00,4.00,R,NaN,NaN


In [159]:
df_ad.isnull().sum()

Customer                        0
SCORE_INDICATOR_CR22          843
SCORE_CR22                    843
SCORE_BAND_CR22               843
SCORECARD_CR21                819
NO_SCORE                      819
RISK_RELATIVE_CR21            819
RISK_RELATIVE_ODDS_CR21       819
NO_MONTHS_OLDEST_ACCOUNT        0
OPEN_ACCT_CR22                  0
CLOSED_ACCT_CR22                0
MORTGAGE_CR22                   0
CREDIT_CARD_CR22                0
TELCO_CR22                      0
UTILITIES_CR22                  0
OVERDRAFT_CR22                  0
PERSONAL_LOAN_CR22              0
AUTO_LOAN_CR22                  0
CREDIT_CARD_AMT_CR22            0
OVERDRAFT_AMT_CR22              0
PERSONAL_LOAN_AMT_CR22          0
AUTO_LOAN_AMT_CR22              0
OPEN_ACCT_CR21                  0
CLOSED_ACCT_CR21                0
MORTGAGE_CR21                   0
CREDIT_CARD_CR21                0
TELCO_CR21                      0
UTILITIES_CR21                  0
OVERDRAFT_CR21                  0
PERSONAL_LOAN_

In [160]:
df_ad.columns

Index(['Customer', 'SCORE_INDICATOR_CR22', 'SCORE_CR22', 'SCORE_BAND_CR22',
       'SCORECARD_CR21', 'NO_SCORE', 'RISK_RELATIVE_CR21',
       'RISK_RELATIVE_ODDS_CR21', 'NO_MONTHS_OLDEST_ACCOUNT', 'OPEN_ACCT_CR22',
       'CLOSED_ACCT_CR22', 'MORTGAGE_CR22', 'CREDIT_CARD_CR22', 'TELCO_CR22',
       'UTILITIES_CR22', 'OVERDRAFT_CR22', 'PERSONAL_LOAN_CR22',
       'AUTO_LOAN_CR22', 'CREDIT_CARD_AMT_CR22', 'OVERDRAFT_AMT_CR22',
       'PERSONAL_LOAN_AMT_CR22', 'AUTO_LOAN_AMT_CR22', 'OPEN_ACCT_CR21',
       'CLOSED_ACCT_CR21', 'MORTGAGE_CR21', 'CREDIT_CARD_CR21', 'TELCO_CR21',
       'UTILITIES_CR21', 'OVERDRAFT_CR21', 'PERSONAL_LOAN_CR21',
       'AUTO_LOAN_CR21', 'CREDIT_CARD_AMT_CR21', 'OVERDRAFT_AMT_CR21',
       'PERSONAL_LOAN_AMT_CR21', 'AUTO_LOAN_AMT_CR21', 'ENQUIRY_CNT_CR22',
       'ENQUIRY_CNT_1M_CR22', 'ENQUIRY_CNT_3M_CR22', 'ENQUIRY_CNT_6M_CR22',
       'ENQUIRY_CNT_12M_CR22', 'ENQUIRY_CNT_CR21', 'ENQUIRY_CNT_1M_CR21',
       'ENQUIRY_CNT_3M_CR21', 'ENQUIRY_CNT_6M_CR21', 'ENQUI

### **Finding Extra Customer Available in **Additional Information** Data sets**

In [161]:
extra_in_ad = set(df_ad['Customer']) - set(df_data['Customer'])
len(extra_in_ad)


12611

**There are 12611 extra customer in Additonal Information**

**Finding Duplicates in Additional Information**

In [162]:
df_ad['Customer'].duplicated().sum()


np.int64(3644)

**In that Additional Information there are 3644 Duplicates Customer**

In [163]:
df_ad_unique = df_ad.drop_duplicates(subset=['Customer'])
df_ad_unique.head()

,Customer,SCORE_INDICATOR_CR22,SCORE_CR22,SCORE_BAND_CR22,SCORECARD_CR21,NO_SCORE,RISK_RELATIVE_CR21,RISK_RELATIVE_ODDS_CR21,NO_MONTHS_OLDEST_ACCOUNT,OPEN_ACCT_CR22,CLOSED_ACCT_CR22,MORTGAGE_CR22,CREDIT_CARD_CR22,TELCO_CR22,UTILITIES_CR22,OVERDRAFT_CR22,PERSONAL_LOAN_CR22,AUTO_LOAN_CR22,CREDIT_CARD_AMT_CR22,OVERDRAFT_AMT_CR22,PERSONAL_LOAN_AMT_CR22,AUTO_LOAN_AMT_CR22,OPEN_ACCT_CR21,CLOSED_ACCT_CR21,MORTGAGE_CR21,CREDIT_CARD_CR21,TELCO_CR21,UTILITIES_CR21,OVERDRAFT_CR21,PERSONAL_LOAN_CR21,AUTO_LOAN_CR21,CREDIT_CARD_AMT_CR21,OVERDRAFT_AMT_CR21,PERSONAL_LOAN_AMT_CR21,AUTO_LOAN_AMT_CR21,ENQUIRY_CNT_CR22,ENQUIRY_CNT_1M_CR22,ENQUIRY_CNT_3M_CR22,ENQUIRY_CNT_6M_CR22,ENQUIRY_CNT_12M_CR22,ENQUIRY_CNT_CR21,ENQUIRY_CNT_1M_CR21,ENQUIRY_CNT_3M_CR21,ENQUIRY_CNT_6M_CR21,ENQUIRY_CNT_12M_CR21,ENQUIRY_MAX_AMT_1M,ENQUIRY_MAX_AMT_3M,ENQUIRY_MAX_AMT_6M,ENQUIRY_MAX_AMT_12M,DEFAULT_CNT_CR22,DEFAULT_OPEN_CNT_CR22,DEFAULT_CNT_3M_CR22,DEFAULT_CNT_6M_CR22,DEFAULT_CNT_12M_CR22,DEFAULT_UNPAID_AMT_CR22,DEFAULT_CNT_CR21,DEFAULT_OPEN_CNT_CR21,DEFAULT_UNPAID_AMT_CR21,DEFAULT_CNT_3M_CR21,DEFAULT_CNT_6M_CR21,DEFAULT_CNT_12M_CR21,WORST_1M,WORST_3M,WORST_6M,WORST_12M,WORST_18M,WORST_24M,HAD_SERVICES,DOC_TYPE,VISA_TYPE,EMPLOYED_STATUS,EMPLOYED_TYPE,EMPLOYED_NO_MTHS,NO_ADDRESS_AT_MONTHS,CD_STS_RESIDENTIAL,AGENCY_AMT,MTH_TILL_AGENCY
0,2984743758,Clean Positive,373.00,H,CTS1A,608.00,4.09,16.91,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,Intl Passport and Visa,Work Visa,E,F,3.00,0.00,R,NaN,NaN
1,2984205798,Clean Positive,1024.00,A,CTSSH,827.00,18.66,77.17,268,2,0,0,2,0,0,0,0,0,18000,0,0,0,3,0,0,0,1,0,0,0,0,0,0,0,0,2,1,1,1,1,2,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,AU Driver Licence,Resident,R,U,0.00,120.00,O,NaN,NaN
2,2983553714,Clean Positive,1005.00,A,CTSSH,833.00,19.45,80.45,339,6,1,5,1,0,0,0,0,0,3000,0,0,0,6,1,5,0,0,0,0,0,0,0,0,0,0,2,1,1,1,1,3,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,AU Driver Licence,Resident,E,F,131.00,24.00,O,NaN,NaN
3,2972077238,Clean Positive,506.00,G,CTS1A,724.00,9.14,37.79,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,AU Driver Licence,Resident,E,P,2.00,2.00,R,NaN,NaN
4,2975514910,Dirty Positive,98.00,H,CTSDP,484.00,1.73,7.16,4,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6,1,1,2,5,7,1,1,4,5,0,0,529,4500,1,1,0,0,0,726,0,0,0,0,0,0,3,3,3,3,3,3,1,AU Driver Licence,Resident,E,F,20.00,4.00,R,NaN,NaN


**After Finding Unique Customer in Addtional Information, There are 101328 customer**

**Merging Data Table and Additonal Information after removing extra Customers and Duplicates as Credit Risk Master Table**

In [164]:
df_crm = pd.merge(df_data, df_ad_unique, how="left", on=["Customer","CD_STS_RESIDENTIAL","NO_SCORE"])
df_crm.shape


(91849, 97)

<div style="text-align:center;">
  <span style="color:Green; font-size:30px; font-weight:bold;">
    Credit Risk Master  Table
  </span>
</div>


In [165]:
df_crm.head()

,OUTCOME,Customer,RISK_SCORE,BUREAU_ENQUIRIES_12_MONTHS,LICENCE_AVAILABLE,RESIDENTIAL,BUREAU_DEFAULT,CD_OCCUPATION,CD_STS_RESIDENTIAL,DEFAULT_AMT,ENQUIRIES_6M,ENQUIRIES_12M,DEROGATORIES,DEFAULTS_12M,NO_MONTHS_VCR221,NO_STATUS_VCR183,CD_STS_VCR184,NO_MONTHS_VCR228,NO_MONTHS_VCR226,NO_SUM_VCR053,NO_MONTHS_VCR258,SCORECARD,NO_SCORE,SCORE_INDICATOR_CR22,SCORE_CR22,SCORE_BAND_CR22,SCORECARD_CR21,RISK_RELATIVE_CR21,RISK_RELATIVE_ODDS_CR21,NO_MONTHS_OLDEST_ACCOUNT,OPEN_ACCT_CR22,CLOSED_ACCT_CR22,MORTGAGE_CR22,CREDIT_CARD_CR22,TELCO_CR22,UTILITIES_CR22,OVERDRAFT_CR22,PERSONAL_LOAN_CR22,AUTO_LOAN_CR22,CREDIT_CARD_AMT_CR22,OVERDRAFT_AMT_CR22,PERSONAL_LOAN_AMT_CR22,AUTO_LOAN_AMT_CR22,OPEN_ACCT_CR21,CLOSED_ACCT_CR21,MORTGAGE_CR21,CREDIT_CARD_CR21,TELCO_CR21,UTILITIES_CR21,OVERDRAFT_CR21,PERSONAL_LOAN_CR21,AUTO_LOAN_CR21,CREDIT_CARD_AMT_CR21,OVERDRAFT_AMT_CR21,PERSONAL_LOAN_AMT_CR21,AUTO_LOAN_AMT_CR21,ENQUIRY_CNT_CR22,ENQUIRY_CNT_1M_CR22,ENQUIRY_CNT_3M_CR22,ENQUIRY_CNT_6M_CR22,ENQUIRY_CNT_12M_CR22,ENQUIRY_CNT_CR21,ENQUIRY_CNT_1M_CR21,ENQUIRY_CNT_3M_CR21,ENQUIRY_CNT_6M_CR21,ENQUIRY_CNT_12M_CR21,ENQUIRY_MAX_AMT_1M,ENQUIRY_MAX_AMT_3M,ENQUIRY_MAX_AMT_6M,ENQUIRY_MAX_AMT_12M,DEFAULT_CNT_CR22,DEFAULT_OPEN_CNT_CR22,DEFAULT_CNT_3M_CR22,DEFAULT_CNT_6M_CR22,DEFAULT_CNT_12M_CR22,DEFAULT_UNPAID_AMT_CR22,DEFAULT_CNT_CR21,DEFAULT_OPEN_CNT_CR21,DEFAULT_UNPAID_AMT_CR21,DEFAULT_CNT_3M_CR21,DEFAULT_CNT_6M_CR21,DEFAULT_CNT_12M_CR21,WORST_1M,WORST_3M,WORST_6M,WORST_12M,WORST_18M,WORST_24M,HAD_SERVICES,DOC_TYPE,VISA_TYPE,EMPLOYED_STATUS,EMPLOYED_TYPE,EMPLOYED_NO_MTHS,NO_ADDRESS_AT_MONTHS,AGENCY_AMT,MTH_TILL_AGENCY
0,Good,2975536526,628,queries 12 Months 1-2,Driver Licence Check Missing,Rented,Default Current- Missing,E,R,0,2,2,0,0,0,-,-,0,0,0,0,TAR1A,724,Clean Positive,586.00,F,CTS1A,9.14,37.79,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,Intl Passport and Visa,Work Visa,E,F,1.00,1.00,NaN,NaN
1,Good,2999806086,813,queries 12 Months 1-2,Driver Licence Check Yes,Rented,Default Current- Missing,E,R,0,2,2,0,0,0,1,-,24,0,4000,23,TAR1A,775,Clean Positive,636.00,D,CTS1A,13.01,53.82,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,1.00,1.00,1.00,1.00,AU Driver Licence,Resident,E,F,86.00,49.00,NaN,NaN
2,Good,2800835182,1012,queries 12 Months 4-5,Driver Licence Check Yes,Owned,Default Current- Missing,E,O,0,5,5,0,0,0,0,-,24,0,380277,24,SFJR1A,815,Clean Positive,952.00,A,CTSSH,17.17,71.01,110.00,7.00,2.00,6.00,1.00,0.00,0.00,0.00,0.00,0.00,17500.00,0.00,0.00,0.00,8.00,1.00,6.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,3.00,2.00,2.00,3.00,3.00,6.00,1.00,1.00,2.00,2.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,AU Driver Licence,Resident,E,F,61.00,129.00,NaN,NaN
3,Good,2975640662,415,queries 12 Months 4-5,Driver Licence Check Missing,Rented,Default Current1-1000,E,R,459,2,4,1,0,0,-,-,0,0,0,0,HSHSOL,328,Clean Positive,534.00,F,CTSAD,-1.70,2.43,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,1.00,1.00,1.00,1.00,4.00,1.00,1.00,1.00,3.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,1.00,459.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,Intl Passport and Visa,Resident,E,F,15.00,60.00,NaN,NaN
4,Good,2979232086,556,queries 12 Months 1-2,Driver Licence Check Missing,Board or Living With Relative,Default Current- Missing,U,B,0,2,2,0,0,0,-,-,0,0,0,0,TAR1A,692,Clean Positive,419.00,H,CTS1A,7.32,30.27,0.00,0.00,0.00,0.00,0

In [166]:
df_crm.columns.unique()

Index(['OUTCOME', 'Customer', 'RISK_SCORE', 'BUREAU_ENQUIRIES_12_MONTHS',
       'LICENCE_AVAILABLE', 'RESIDENTIAL', 'BUREAU_DEFAULT', 'CD_OCCUPATION',
       'CD_STS_RESIDENTIAL', 'DEFAULT_AMT', 'ENQUIRIES_6M', 'ENQUIRIES_12M',
       'DEROGATORIES', 'DEFAULTS_12M', 'NO_MONTHS_VCR221', 'NO_STATUS_VCR183',
       'CD_STS_VCR184', 'NO_MONTHS_VCR228', 'NO_MONTHS_VCR226',
       'NO_SUM_VCR053', 'NO_MONTHS_VCR258', 'SCORECARD', 'NO_SCORE',
       'SCORE_INDICATOR_CR22', 'SCORE_CR22', 'SCORE_BAND_CR22',
       'SCORECARD_CR21', 'RISK_RELATIVE_CR21', 'RISK_RELATIVE_ODDS_CR21',
       'NO_MONTHS_OLDEST_ACCOUNT', 'OPEN_ACCT_CR22', 'CLOSED_ACCT_CR22',
       'MORTGAGE_CR22', 'CREDIT_CARD_CR22', 'TELCO_CR22', 'UTILITIES_CR22',
       'OVERDRAFT_CR22', 'PERSONAL_LOAN_CR22', 'AUTO_LOAN_CR22',
       'CREDIT_CARD_AMT_CR22', 'OVERDRAFT_AMT_CR22', 'PERSONAL_LOAN_AMT_CR22',
       'AUTO_LOAN_AMT_CR22', 'OPEN_ACCT_CR21', 'CLOSED_ACCT_CR21',
       'MORTGAGE_CR21', 'CREDIT_CARD_CR21', 'TELCO_CR21', 'UT

**In credit-related problems, data typically comes from three major areas: Customer data, Loan/Credit data, and Bureau data. After merging these datasets, it is often useful to separate them again into individual tables to make data cleaning easier before performing Exploratory Data Analysis (EDA).**

---------------------------------------------------------------------------------------------------------------

<!DOCTYPE html>
<html>
<head>
<style>
    .data-cleaning-title {
        font-family: Arial, Helvetica, sans-serif;
        font-size: 26px;
        font-weight: bold;
        color: #1f4fd8;
        text-align: center;
        margin: 20px 0;
    }
</style>
</head>

<body>

<div class="data-cleaning-title">Data Cleaning</div>

</body>
</html>


**Data Cleaning Approach**

To make the data cleaning process structured and manageable, I segmented the dataset into three logical parts: Customer, Loan, and Bureau. Each segment has different data characteristics and business rules, so cleaning them separately helped avoid confusion and errors.

- **Customer data** was cleaned for demographic consistency, missing values, and stability-related fields such as employment and address duration.

- **Loan data** was cleaned for loan-specific validations like loan amount, tenure, EMI, and outcome-related fields.

- **Bureau data** was cleaned using credit-risk domain logic, including score bands, default indicators, enquiry behavior, and delinquency status.

After cleaning each segment independently, I merged them into a single modeling dataset as **Credit Risk Master table** and performed **EDA** again to validate relationships and risk patterns.

### **Customer Table**

In [167]:
customer_cols = [
    'OUTCOME',
    'Customer',
    'LICENCE_AVAILABLE',
    'RESIDENTIAL',
    'CD_OCCUPATION',
    'CD_STS_RESIDENTIAL',   
    'HAD_SERVICES',
    'DOC_TYPE',
    'VISA_TYPE',
    'EMPLOYED_STATUS',
    'EMPLOYED_TYPE',
    'EMPLOYED_NO_MTHS',
    'NO_ADDRESS_AT_MONTHS'
]


In [168]:
df_age = pd.read_excel(
    "age_data.xlsx"
)



In [169]:
df_age.head()

,Customer,APPLICANT_AGE
0,2975536526,Age 30-34
1,2999806086,Age 30-34
2,2800835182,Age 35-44
3,2975640662,Redeveloped Age 25-29
4,2979232086,Age 18-24


In [170]:
# Filter only customer data
df_customer = df_crm[customer_cols]


# Display first rows
df_customer.head()

,OUTCOME,Customer,LICENCE_AVAILABLE,RESIDENTIAL,CD_OCCUPATION,CD_STS_RESIDENTIAL,HAD_SERVICES,DOC_TYPE,VISA_TYPE,EMPLOYED_STATUS,EMPLOYED_TYPE,EMPLOYED_NO_MTHS,NO_ADDRESS_AT_MONTHS
0,Good,2975536526,Driver Licence Check Missing,Rented,E,R,1.00,Intl Passport and Visa,Work Visa,E,F,1.00,1.00
1,Good,2999806086,Driver Licence Check Yes,Rented,E,R,1.00,AU Driver Licence,Resident,E,F,86.00,49.00
2,Good,2800835182,Driver Licence Check Yes,Owned,E,O,1.00,AU Driver Licence,Resident,E,F,61.00,129.00
3,Good,2975640662,Driver Licence Check Missing,Rented,E,R,1.00,Intl Passport and Visa,Resident,E,F,15.00,60.00
4,Good,2979232086,Driver Licence Check Missing,Board or Living With Relative,U,B,1.00,Intl Passport and Visa,Work Visa,U,U,0.00,1.00


In [171]:
df_customer = df_customer.merge(df_age, how='left', on='Customer')

In [172]:
df_customer.columns

Index(['OUTCOME', 'Customer', 'LICENCE_AVAILABLE', 'RESIDENTIAL',
       'CD_OCCUPATION', 'CD_STS_RESIDENTIAL', 'HAD_SERVICES', 'DOC_TYPE',
       'VISA_TYPE', 'EMPLOYED_STATUS', 'EMPLOYED_TYPE', 'EMPLOYED_NO_MTHS',
       'NO_ADDRESS_AT_MONTHS', 'APPLICANT_AGE'],
      dtype='object')

In [173]:
df_customer.shape

(91849, 14)

In [174]:
df_customer.dtypes

OUTCOME                  object
Customer                  int64
LICENCE_AVAILABLE        object
RESIDENTIAL              object
CD_OCCUPATION            object
CD_STS_RESIDENTIAL       object
HAD_SERVICES            float64
DOC_TYPE                 object
VISA_TYPE                object
EMPLOYED_STATUS          object
EMPLOYED_TYPE            object
EMPLOYED_NO_MTHS        float64
NO_ADDRESS_AT_MONTHS    float64
APPLICANT_AGE            object
dtype: object

In [175]:
df_customer.isnull().sum()

OUTCOME                     0
Customer                    0
LICENCE_AVAILABLE           0
RESIDENTIAL                 0
CD_OCCUPATION               0
CD_STS_RESIDENTIAL          0
HAD_SERVICES             7425
DOC_TYPE                 8727
VISA_TYPE                8716
EMPLOYED_STATUS         15093
EMPLOYED_TYPE           15093
EMPLOYED_NO_MTHS        15093
NO_ADDRESS_AT_MONTHS     7425
APPLICANT_AGE               0
dtype: int64

In [176]:
df_customer_na = df_customer[df_customer.isna().any(axis=1)]
df_customer_na.head()



,OUTCOME,Customer,LICENCE_AVAILABLE,RESIDENTIAL,CD_OCCUPATION,CD_STS_RESIDENTIAL,HAD_SERVICES,DOC_TYPE,VISA_TYPE,EMPLOYED_STATUS,EMPLOYED_TYPE,EMPLOYED_NO_MTHS,NO_ADDRESS_AT_MONTHS,APPLICANT_AGE
11,Good,2829349414,Driver Licence Check Yes,Rented,E,R,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Age 30-34
13,Good,2992968286,Driver Licence Check Missing,Rented,U,Missing,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Age 30-34
18,Good,761398138,Driver Licence Check Missing,Owned,S,O,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Age 54+
26,Good,2951915674,Driver Licence Check Yes,Rented,E,R,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Age 18-24
28,Bad,2978936566,Driver Licence Check Yes,Rented,Missing,R,1.00,AU Driver Licence,Resident,NaN,NaN,NaN,0.00,Age 35-44


In [177]:
df_customer.duplicated().sum()

np.int64(0)

In [178]:
df_customer.nunique()

OUTCOME                     3
Customer                91849
LICENCE_AVAILABLE           3
RESIDENTIAL                 5
CD_OCCUPATION              10
CD_STS_RESIDENTIAL          7
HAD_SERVICES                2
DOC_TYPE                    8
VISA_TYPE                   3
EMPLOYED_STATUS             9
EMPLOYED_TYPE               5
EMPLOYED_NO_MTHS          363
NO_ADDRESS_AT_MONTHS      432
APPLICANT_AGE               6
dtype: int64

In [179]:
for col in df_customer.select_dtypes(include='object').columns:
    print(col, "-->", df_customer[col].unique())


OUTCOME --> ['Good' 'Bad' 'Rejected - Policy Decline']
LICENCE_AVAILABLE --> ['Driver Licence Check Missing' 'Driver Licence Check Yes'
 'Driver Licence Check No']
RESIDENTIAL --> ['Rented' 'Owned' 'Board or Living With Relative' 'Living At Home'
 'Residential Missing Not Willing To Supply']
CD_OCCUPATION --> ['E' 'U' 'P' 'R' 'S' 'Missing' 'B' 'Z' 'T' 'H']
CD_STS_RESIDENTIAL --> ['R' 'O' 'B' 'Missing' 'L' 'W' 'Z']
DOC_TYPE --> ['Intl Passport and Visa' 'AU Driver Licence' 'AU Passport' nan
 'HAAU 18+ Card' 'Australian Passport' 'Fire Arms Licence'
 'Defence Force ID(picture card)' 'AU Birth Certificate']
VISA_TYPE --> ['Work Visa' 'Resident' 'Student' nan]
EMPLOYED_STATUS --> ['E' 'U' 'P' nan 'R' 'B' 'Z' 'S' 'T' 'H']
EMPLOYED_TYPE --> ['F' 'U' nan 'P' 'S' 'H']
APPLICANT_AGE --> ['Age 30-34' 'Age 35-44' 'Redeveloped Age 25-29' 'Age 18-24' 'Age 54+'
 'Age 45-54']


In [180]:
applicant_age_map = {
    'Age 30-34': "30-34",
    'Age 35-44' : "35-44",
    'Redeveloped Age 25-29' : "25 - 29",
    'Age 18-24': "18-24",
    'Age 54+' : "54+",
    'Age 45-54': "45-54"
}

df_customer["APPLICANT_AGE"] = df_customer["APPLICANT_AGE"].map(applicant_age_map)

license_available_map = {
    'Driver Licence Check Missing': "Missing",
    'Driver Licence Check Yes': "yes",
    'Driver Licence Check No': "no"

}

df_customer["LICENCE_AVAILABLE"] = df_customer["LICENCE_AVAILABLE"].map(license_available_map )



residential_map = {
    "Owned": "Owned",
    "Rented": "Rented",
    "Living At Home": "Living_With_Family",
    "Board or Living With Relative": "Living_With_Family",
    "Residential Missing Not Willing To Supply": "Missing"
}

df_customer["RESIDENTIAL"] = df_customer["RESIDENTIAL"].map(residential_map)

occupational_map ={
    'E':"employed",
    'U':"unemployed",
    'P':"student",
    'R':"retired",
    'S':"self_employed",
    'B':"benefits",
    'Z':"unwilling to supply",
    'T':"superannuation",
    'H':"home_excecutive"
}

df_customer["CD_OCCUPATION"] = df_customer["CD_OCCUPATION"].map(occupational_map)
df_customer["EMPLOYED_STATUS"] = df_customer["EMPLOYED_STATUS"].map(occupational_map)
df_customer["EMPLOYED_TYPE"] = df_customer["EMPLOYED_TYPE"].map(occupational_map)

In [181]:
df_customer["APPLICANT_AGE_AVG"] = (
    df_customer["APPLICANT_AGE"]
    .astype(str)
    .str.replace(" ", "")
    .str.extract(r"(\d+)-(\d+)")
    .astype(float)
    .mean(axis=1)
)


In [182]:
df_customer["APPLICANT_AGE_AVG"] = df_customer["APPLICANT_AGE_AVG"].round().astype("Int64")


In [183]:
df_customer.head()

,OUTCOME,Customer,LICENCE_AVAILABLE,RESIDENTIAL,CD_OCCUPATION,CD_STS_RESIDENTIAL,HAD_SERVICES,DOC_TYPE,VISA_TYPE,EMPLOYED_STATUS,EMPLOYED_TYPE,EMPLOYED_NO_MTHS,NO_ADDRESS_AT_MONTHS,APPLICANT_AGE,APPLICANT_AGE_AVG
0,Good,2975536526,Missing,Rented,employed,R,1.00,Intl Passport and Visa,Work Visa,employed,NaN,1.00,1.00,30-34,32
1,Good,2999806086,yes,Rented,employed,R,1.00,AU Driver Licence,Resident,employed,NaN,86.00,49.00,30-34,32
2,Good,2800835182,yes,Owned,employed,O,1.00,AU Driver Licence,Resident,employed,NaN,61.00,129.00,35-44,40
3,Good,2975640662,Missing,Rented,employed,R,1.00,Intl Passport and Visa,Resident,employed,NaN,15.00,60.00,25 - 29,27
4,Good,2979232086,Missing,Living_With_Family,unemployed,B,1.00,Intl Passport and Visa,Work Visa,unemployed,unemployed,0.00,1.00,18-24,21


In [184]:
df_customer.dtypes

OUTCOME                  object
Customer                  int64
LICENCE_AVAILABLE        object
RESIDENTIAL              object
CD_OCCUPATION            object
CD_STS_RESIDENTIAL       object
HAD_SERVICES            float64
DOC_TYPE                 object
VISA_TYPE                object
EMPLOYED_STATUS          object
EMPLOYED_TYPE            object
EMPLOYED_NO_MTHS        float64
NO_ADDRESS_AT_MONTHS    float64
APPLICANT_AGE            object
APPLICANT_AGE_AVG         Int64
dtype: object

In [185]:
for col in df_customer.select_dtypes(include='object').columns:
    print(col, "-->", df_customer[col].unique())

OUTCOME --> ['Good' 'Bad' 'Rejected - Policy Decline']
LICENCE_AVAILABLE --> ['Missing' 'yes' 'no']
RESIDENTIAL --> ['Rented' 'Owned' 'Living_With_Family' 'Missing']
CD_OCCUPATION --> ['employed' 'unemployed' 'student' 'retired' 'self_employed' nan
 'benefits' 'unwilling to supply' 'superannuation' 'home_excecutive']
CD_STS_RESIDENTIAL --> ['R' 'O' 'B' 'Missing' 'L' 'W' 'Z']
DOC_TYPE --> ['Intl Passport and Visa' 'AU Driver Licence' 'AU Passport' nan
 'HAAU 18+ Card' 'Australian Passport' 'Fire Arms Licence'
 'Defence Force ID(picture card)' 'AU Birth Certificate']
VISA_TYPE --> ['Work Visa' 'Resident' 'Student' nan]
EMPLOYED_STATUS --> ['employed' 'unemployed' 'student' nan 'retired' 'benefits'
 'unwilling to supply' 'self_employed' 'superannuation' 'home_excecutive']
EMPLOYED_TYPE --> [nan 'unemployed' 'student' 'self_employed' 'home_excecutive']
APPLICANT_AGE --> ['30-34' '35-44' '25 - 29' '18-24' '54+' '45-54']


In [186]:
for col in df_customer.select_dtypes(include='object').columns:
    df_customer[col] = df_customer[col].fillna("Missing")


In [187]:
for col in df_customer.select_dtypes(include='object').columns:
    print(col, "-->", df_customer[col].unique())

OUTCOME --> ['Good' 'Bad' 'Rejected - Policy Decline']
LICENCE_AVAILABLE --> ['Missing' 'yes' 'no']
RESIDENTIAL --> ['Rented' 'Owned' 'Living_With_Family' 'Missing']
CD_OCCUPATION --> ['employed' 'unemployed' 'student' 'retired' 'self_employed' 'Missing'
 'benefits' 'unwilling to supply' 'superannuation' 'home_excecutive']
CD_STS_RESIDENTIAL --> ['R' 'O' 'B' 'Missing' 'L' 'W' 'Z']
DOC_TYPE --> ['Intl Passport and Visa' 'AU Driver Licence' 'AU Passport' 'Missing'
 'HAAU 18+ Card' 'Australian Passport' 'Fire Arms Licence'
 'Defence Force ID(picture card)' 'AU Birth Certificate']
VISA_TYPE --> ['Work Visa' 'Resident' 'Student' 'Missing']
EMPLOYED_STATUS --> ['employed' 'unemployed' 'student' 'Missing' 'retired' 'benefits'
 'unwilling to supply' 'self_employed' 'superannuation' 'home_excecutive']
EMPLOYED_TYPE --> ['Missing' 'unemployed' 'student' 'self_employed' 'home_excecutive']
APPLICANT_AGE --> ['30-34' '35-44' '25 - 29' '18-24' '54+' '45-54']


In [188]:
df_customer.isna().sum()

OUTCOME                     0
Customer                    0
LICENCE_AVAILABLE           0
RESIDENTIAL                 0
CD_OCCUPATION               0
CD_STS_RESIDENTIAL          0
HAD_SERVICES             7425
DOC_TYPE                    0
VISA_TYPE                   0
EMPLOYED_STATUS             0
EMPLOYED_TYPE               0
EMPLOYED_NO_MTHS        15093
NO_ADDRESS_AT_MONTHS     7425
APPLICANT_AGE               0
APPLICANT_AGE_AVG       17592
dtype: int64

**When employment status is missing, we interpret it as no active employment and set employment duration to zero to preserve logical consistency and risk signal**

In [189]:
df_employed_status_2 = df_customer.groupby("EMPLOYED_STATUS")["EMPLOYED_NO_MTHS"].mean()
df_employed_status_2

EMPLOYED_STATUS
Missing                 NaN
benefits               1.66
employed              32.01
home_excecutive        2.56
retired                3.39
self_employed         34.51
student                0.76
superannuation         3.08
unemployed             0.22
unwilling to supply    0.31
Name: EMPLOYED_NO_MTHS, dtype: float64

In [190]:
df_customer["EMPLOYED_NO_MTHS"] = df_customer["EMPLOYED_NO_MTHS"].fillna(0)


In [191]:
df_customer.isna().sum()

OUTCOME                     0
Customer                    0
LICENCE_AVAILABLE           0
RESIDENTIAL                 0
CD_OCCUPATION               0
CD_STS_RESIDENTIAL          0
HAD_SERVICES             7425
DOC_TYPE                    0
VISA_TYPE                   0
EMPLOYED_STATUS             0
EMPLOYED_TYPE               0
EMPLOYED_NO_MTHS            0
NO_ADDRESS_AT_MONTHS     7425
APPLICANT_AGE               0
APPLICANT_AGE_AVG       17592
dtype: int64

In [192]:
df_employed_status_2 = df_customer.groupby("RESIDENTIAL")["NO_ADDRESS_AT_MONTHS"].mean()
df_employed_status_2

RESIDENTIAL
Living_With_Family   41.21
Missing              68.41
Owned                69.33
Rented               17.36
Name: NO_ADDRESS_AT_MONTHS, dtype: float64

-----------------------------------------------------------------------------------------------

## **Loan/Credit data**

In [193]:
credit_risk_cols = [
    'Customer',
    'RISK_SCORE',
    'BUREAU_DEFAULT',
    'DEFAULT_AMT',
    'DEFAULTS_12M',
    'DEROGATORIES',
    'NO_SCORE',
    'SCORECARD',
    'SCORE_INDICATOR_CR22',
    'SCORE_CR22',
    'SCORE_BAND_CR22',
    'SCORECARD_CR21',
    'RISK_RELATIVE_CR21',
    'RISK_RELATIVE_ODDS_CR21',

    # Delinquency-related VCR variables
    'NO_MONTHS_VCR221',
    'NO_STATUS_VCR183',
    'CD_STS_VCR184',
    'NO_MONTHS_VCR228',
    'NO_MONTHS_VCR226',
    'NO_SUM_VCR053',
    'NO_MONTHS_VCR258',

    # Worst payment behaviour
    'WORST_1M', 'WORST_3M', 'WORST_6M',
    'WORST_12M', 'WORST_18M', 'WORST_24M'
]


df_loan = df_crm[credit_risk_cols]

df_loan.head()

,Customer,RISK_SCORE,BUREAU_DEFAULT,DEFAULT_AMT,DEFAULTS_12M,DEROGATORIES,NO_SCORE,SCORECARD,SCORE_INDICATOR_CR22,SCORE_CR22,SCORE_BAND_CR22,SCORECARD_CR21,RISK_RELATIVE_CR21,RISK_RELATIVE_ODDS_CR21,NO_MONTHS_VCR221,NO_STATUS_VCR183,CD_STS_VCR184,NO_MONTHS_VCR228,NO_MONTHS_VCR226,NO_SUM_VCR053,NO_MONTHS_VCR258,WORST_1M,WORST_3M,WORST_6M,WORST_12M,WORST_18M,WORST_24M
0,2975536526,628,Default Current- Missing,0,0,0,724,TAR1A,Clean Positive,586.00,F,CTS1A,9.14,37.79,0,-,-,0,0,0,0,0.00,0.00,0.00,0.00,0.00,0.00
1,2999806086,813,Default Current- Missing,0,0,0,775,TAR1A,Clean Positive,636.00,D,CTS1A,13.01,53.82,0,1,-,24,0,4000,23,0.00,0.00,1.00,1.00,1.00,1.00
2,2800835182,1012,Default Current- Missing,0,0,0,815,SFJR1A,Clean Positive,952.00,A,CTSSH,17.17,71.01,0,0,-,24,0,380277,24,0.00,0.00,0.00,0.00,0.00,0.00
3,2975640662,415,Default Current1-1000,459,0,1,328,HSHSOL,Clean Positive,534.00,F,CTSAD,-1.70,2.43,0,-,-,0,0,0,0,0.00,0.00,0.00,0.00,0.00,0.00
4,2979232086,556,Default Current- Missing,0,0,0,692,TAR1A,Clean Positive,419.00,H,CTS1A,7.32,30.27,0,-,-,0,0,0,0,0.00,0.00,0.00,0.00,0.00,0.00


In [194]:
df_loan.describe()

,Customer,RISK_SCORE,DEFAULT_AMT,DEFAULTS_12M,DEROGATORIES,NO_SCORE,SCORE_CR22,RISK_RELATIVE_CR21,RISK_RELATIVE_ODDS_CR21,NO_MONTHS_VCR221,NO_MONTHS_VCR228,NO_MONTHS_VCR226,NO_SUM_VCR053,NO_MONTHS_VCR258,WORST_1M,WORST_3M,WORST_6M,WORST_12M,WORST_18M,WORST_24M
count,91849.00,91849.00,91849.00,91849.00,91849.00,91849.00,83851.00,84424.00,84424.00,91849.00,91849.00,91849.00,91849.00,91849.00,84424.00,84424.00,84424.00,84424.00,84424.00,84424.00
mean,2885627227.87,721.31,251.49,0.01,0.26,674.25,591.28,7.17,40.30,0.18,10.29,0.34,82132.67,6.65,0.17,0.23,0.28,0.36,0.45,0.52
std,431320891.09,180.84,4393.88,0.11,1.41,168.66,189.68,48.79,32.48,1.01,11.21,1.69,247992.69,10.36,0.76,0.87,0.95,1.07,1.20,1.29
min,400031354.00,-125.00,0.00,0.00,0.00,-999.00,-204.00,-999.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
25%,2975014418.00,614.00,0.00,0.00,0.00,625.00,457.00,4.93,20.39,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
50%,2984592182.00,660.00,0.00,0.00,0.00,718.00,558.00,9.14,37.79,0.00,2.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
75%,2995693558.00,848.00,0.00,0.00,0.00,749.00,733.00,10.87,44.94,0.00,24.00,0.00,15020.00,20.00,0.00,0.00,0.00,0.00,0.00,0.00
max,3006334582.00,1157.00,1017272.00,8.00,253.00,977.00,1203.00,52.77,218.28,12.00,24.00,24.00,8662123.00,24.00,8.00,8.00,8.00,8.00,8.00,8.00


In [195]:
df_loan.shape

(91849, 27)

In [196]:
df_loan.dtypes

Customer                     int64
RISK_SCORE                   int64
BUREAU_DEFAULT              object
DEFAULT_AMT                  int64
DEFAULTS_12M                 int64
DEROGATORIES                 int64
NO_SCORE                     int64
SCORECARD                   object
SCORE_INDICATOR_CR22        object
SCORE_CR22                 float64
SCORE_BAND_CR22             object
SCORECARD_CR21              object
RISK_RELATIVE_CR21         float64
RISK_RELATIVE_ODDS_CR21    float64
NO_MONTHS_VCR221             int64
NO_STATUS_VCR183            object
CD_STS_VCR184               object
NO_MONTHS_VCR228             int64
NO_MONTHS_VCR226             int64
NO_SUM_VCR053                int64
NO_MONTHS_VCR258             int64
WORST_1M                   float64
WORST_3M                   float64
WORST_6M                   float64
WORST_12M                  float64
WORST_18M                  float64
WORST_24M                  float64
dtype: object

**Worst payment behaviour variables were excluded because they represent post-loan information and would introduce data leakage in an EMI approval model.**

In [197]:
worst_cols = [
    'WORST_1M', 'WORST_3M', 'WORST_6M',
    'WORST_12M', 'WORST_18M', 'WORST_24M'
]

df_loan = df_loan.drop(columns=worst_cols)


**I removed SCORE_INDICATOR_CR22 because it showed non-monotonic behavior, where higher scores were associated with higher default risk, violating credit-risk principles.
Since it lacked business interpretability and stability, it was excluded to avoid misleading the model.**

In [198]:
df_loan = df_loan.drop(columns="SCORE_INDICATOR_CR22")

In [199]:
for col in df_loan.select_dtypes(include='object').columns:
    print(col, "-->", df_loan[col].unique())

BUREAU_DEFAULT --> ['Default Current- Missing' 'Default Current1-1000' 'Default Current1000+']
SCORECARD --> ['TAR1A' 'SFJR1A' 'HSHSOL' 'CTSDP' 'INSLV']
SCORE_BAND_CR22 --> ['F' 'D' 'A' 'H' 'G' nan 'C' 'I' 'E' 'B']
SCORECARD_CR21 --> ['CTS1A' 'CTSSH' 'CTSAD' nan 'CTSDP' 'INSLV']
NO_STATUS_VCR183 --> ['-' 1 0 5 'N' 'U' 2 4 6 3 'H' 'X']
CD_STS_VCR184 --> ['-' 0 2 'N' 1 4 6 'X' 3 5 'H']


In [200]:
score_band_cr22_map = {
    'A': 'Low',
    'B': 'Low',
    'C': 'Medium',
    'D': 'Medium',
    'E': 'Medium',
    'F': 'High',
    'G': 'High',
    'H': 'High',
    'I': 'High'
}

df_loan['SCORE_BAND_CR22'] = (
    df_loan['SCORE_BAND_CR22']
    .map(score_band_cr22_map)
    .fillna('No_Score')
)


no_status_vcr183_map = {
    'N': 0,
    0: 0,
    1: 1,
    2: 2,
    3: 3,
    4: 4,
    5: 5,
    6: 6,
    'H': 6,
    'U': -1,
    'X': -1,
    '-': -1
}


df_loan['NO_STATUS_VCR183'] = (
    df_loan['NO_STATUS_VCR183']
    .map(no_status_vcr183_map)
    .fillna('Missing')
)


cd_sts_vcr184_map = {
     'N': 0,
    0: 0,
    1: 1,
    2: 2,
    3: 3,
    4: 4,
    5: 5,
    6: 6,
    'H': 6,
    'U': -1,
    'X': -1,
    '-': -1
}

df_loan['CD_STS_VCR184'] = (
    df_loan['CD_STS_VCR184']
    .map(cd_sts_vcr184_map)
    .fillna('Missing')
)


In [201]:
BUREAU_DEFAULT_map = {
    'Default Current- Missing': "Missing",
    'Default Current1-1000': "1-1000",
    'Default Current1000+': "1000+"


}

df_loan["BUREAU_DEFAULT"] = df_loan["BUREAU_DEFAULT"].map(BUREAU_DEFAULT_map)

In [202]:
df_loan.isnull().sum()

Customer                      0
RISK_SCORE                    0
BUREAU_DEFAULT                0
DEFAULT_AMT                   0
DEFAULTS_12M                  0
DEROGATORIES                  0
NO_SCORE                      0
SCORECARD                     0
SCORE_CR22                 7998
SCORE_BAND_CR22               0
SCORECARD_CR21             7425
RISK_RELATIVE_CR21         7425
RISK_RELATIVE_ODDS_CR21    7425
NO_MONTHS_VCR221              0
NO_STATUS_VCR183              0
CD_STS_VCR184                 0
NO_MONTHS_VCR228              0
NO_MONTHS_VCR226              0
NO_SUM_VCR053                 0
NO_MONTHS_VCR258              0
dtype: int64

In [203]:
for col in df_loan.select_dtypes(include='int64').columns:
    print(col, "-->", df_loan[col].unique())

Customer --> [2975536526 2999806086 2800835182 ... 2971478930 2974937898 2977362690]
RISK_SCORE --> [ 628  813 1012 ... -125   80  136]
DEFAULT_AMT --> [    0   459   215 ... 44218   625  5632]
DEFAULTS_12M --> [0 1 3 2 8 4]
DEROGATORIES --> [  0   1   4   2   8   3   7   6   5  13  11  10  15  21  12  14   9  16
  17  20  18  22  23 253  31  19  27]
NO_SCORE --> [ 724  775  815 ... -216  -33  -35]
NO_MONTHS_VCR221 --> [ 0  4  1  6  3  2 11  5  8  7 12 10  9]
NO_STATUS_VCR183 --> [-1  1  0  5  2  4  6  3]
CD_STS_VCR184 --> [-1  0  2  1  4  6  3  5]
NO_MONTHS_VCR228 --> [ 0 24 18 10 22 17  4 23 21 12 11  3  1 20 19  6 15  8  9  2  5  7 14 16
 13]
NO_MONTHS_VCR226 --> [ 0  4  1  3  8  5  6 13 10  2 11  9 12 14  7 16 18 24 15 19 23 21 17 22
 20]
NO_SUM_VCR053 --> [     0   4000 380277 ... 418677   8933 503009]
NO_MONTHS_VCR258 --> [ 0 23 24 18  7 19 22 17 13 16 15 11 10  3  2  6  8 21  9 20  5 12 14  4
  1]


In [204]:
for col in df_loan.select_dtypes(include='object').columns:
    print(col, "-->", df_loan[col].unique())

BUREAU_DEFAULT --> ['Missing' '1-1000' '1000+']
SCORECARD --> ['TAR1A' 'SFJR1A' 'HSHSOL' 'CTSDP' 'INSLV']
SCORE_BAND_CR22 --> ['High' 'Medium' 'Low' 'No_Score']
SCORECARD_CR21 --> ['CTS1A' 'CTSSH' 'CTSAD' nan 'CTSDP' 'INSLV']


In [205]:
df_loan.duplicated().sum()

np.int64(0)

In [206]:
df_loan= df_loan.rename(columns={
    "NO_MONTHS_VCR221": "Late_Payment_30DPD_Last_12M",
    "NO_STATUS_VCR183": "Credit_Card_Payment_Failure_Count",
    "CD_STS_VCR184": "Personal_Loan_Payment_Failure_Count",
    "NO_MONTHS_VCR228": "Recent_Payment_Irregularity_Flag",
    "NO_MONTHS_VCR226":"Late_Payment_30DPD_Last_24M",
    "NO_SUM_VCR053": "Cumulative_Delinquency_Exposure",
    "NO_MONTHS_VCR258": "Long_Term_Payment_Delinquency_Count"

})  


#### **Analysing Patten Between Score_band_CR22 vs Score_Cr22** 

In [207]:
grouped_df = (
    df_loan
    .groupby(['SCORE_BAND_CR22',	'SCORE_CR22' ])
    .size()
    .reset_index(name='COUNT')
)

grouped_df.head()

,SCORE_BAND_CR22,SCORE_CR22,COUNT
0,High,-204.00,1
1,High,-94.00,1
2,High,-87.00,1
3,High,-78.00,1
4,High,-68.00,1


In [208]:
bins = [-999, 500, 607, 715, 1203]
labels = ['Very_High_Risk', 'High', 'Medium', 'Low']

range_summary = (
    df_loan
    .assign(
        CR22_RANGE=pd.cut(df_loan['SCORE_CR22'], bins=bins, labels=labels)
    )
    .groupby('CR22_RANGE', as_index=False)
    .agg(
        MIN_SCORE=('SCORE_CR22', 'min'),
        MAX_SCORE=('SCORE_CR22', 'max')
    )
)

range_summary


C:\Users\sree varshan\AppData\Local\Temp\ipykernel_23720\3598300506.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  .groupby('CR22_RANGE', as_index=False)


,CR22_RANGE,MIN_SCORE,MAX_SCORE
0,Very_High_Risk,-204.00,500.00
1,High,501.00,607.00
2,Medium,608.00,715.00
3,Low,716.00,1203.00


#### **Analysing the Patten Between RISK_SCORE and SCORE_CR22**

In [209]:
grouped_df = (
    df_loan
    .groupby(['RISK_SCORE',	'SCORE_CR22' ])
    .size()
    .reset_index(name='COUNT')
)

grouped_df.head()

,RISK_SCORE,SCORE_CR22,COUNT
0,-76,270.00,1
1,-54,2.00,1
2,-51,1.00,1
3,-17,203.00,1
4,-12,159.00,1


#### **Analysing the patten between RISK_SCORE and SCORE_CR22 and SCORE_BAND_CR22**

In [210]:
grouped_df = (
    df_loan
    .groupby(['RISK_SCORE',	'SCORE_CR22','SCORE_BAND_CR22'])
    .size()
    .reset_index(name='COUNT')
)

grouped_df.head()

,RISK_SCORE,SCORE_CR22,SCORE_BAND_CR22,COUNT
0,-76,270.00,High,1
1,-54,2.00,High,1
2,-51,1.00,High,1
3,-17,203.00,High,1
4,-12,159.00,High,1


-----------------------------------------------

### **Bureau**

In [211]:
bureau_cols = [
    'Customer',
    'BUREAU_ENQUIRIES_12_MONTHS',
    'ENQUIRIES_6M',
    'ENQUIRIES_12M',

    # Bureau enquiry patterns
    #CR22
    'ENQUIRY_CNT_CR22', 'ENQUIRY_CNT_1M_CR22', 'ENQUIRY_CNT_3M_CR22',
    'ENQUIRY_CNT_6M_CR22', 'ENQUIRY_CNT_12M_CR22',
    #CR21
    'ENQUIRY_CNT_CR21', 'ENQUIRY_CNT_1M_CR21', 'ENQUIRY_CNT_3M_CR21',
    'ENQUIRY_CNT_6M_CR21', 'ENQUIRY_CNT_12M_CR21',

    # Enquiry amounts
    'ENQUIRY_MAX_AMT_1M', 'ENQUIRY_MAX_AMT_3M',
    'ENQUIRY_MAX_AMT_6M', 'ENQUIRY_MAX_AMT_12M',

    # CR22 exposure
    'OPEN_ACCT_CR22', 'CLOSED_ACCT_CR22', 'MORTGAGE_CR22',
    'CREDIT_CARD_CR22', 'TELCO_CR22', 'UTILITIES_CR22',
    'OVERDRAFT_CR22', 'PERSONAL_LOAN_CR22', 'AUTO_LOAN_CR22',
    'CREDIT_CARD_AMT_CR22', 'OVERDRAFT_AMT_CR22',
    'PERSONAL_LOAN_AMT_CR22', 'AUTO_LOAN_AMT_CR22',

    # CR21 exposure
    'OPEN_ACCT_CR21', 'CLOSED_ACCT_CR21', 'MORTGAGE_CR21',
    'CREDIT_CARD_CR21', 'TELCO_CR21', 'UTILITIES_CR21',
    'OVERDRAFT_CR21', 'PERSONAL_LOAN_CR21', 'AUTO_LOAN_CR21',
    'CREDIT_CARD_AMT_CR21', 'OVERDRAFT_AMT_CR21',
    'PERSONAL_LOAN_AMT_CR21', 'AUTO_LOAN_AMT_CR21',

    # Bureau defaults
    #CR22
    'DEFAULT_CNT_CR22', 'DEFAULT_OPEN_CNT_CR22',
    'DEFAULT_CNT_3M_CR22', 'DEFAULT_CNT_6M_CR22',
    'DEFAULT_CNT_12M_CR22', 'DEFAULT_UNPAID_AMT_CR22',

    #CR21
    'DEFAULT_CNT_CR21', 'DEFAULT_OPEN_CNT_CR21',
    'DEFAULT_UNPAID_AMT_CR21', 'DEFAULT_CNT_3M_CR21',
    'DEFAULT_CNT_6M_CR21', 'DEFAULT_CNT_12M_CR21',

    # Collection info
    'AGENCY_AMT',
    'MTH_TILL_AGENCY'

  
]


df_bureau = df_crm[bureau_cols]

df_bureau.head()

,Customer,BUREAU_ENQUIRIES_12_MONTHS,ENQUIRIES_6M,ENQUIRIES_12M,ENQUIRY_CNT_CR22,ENQUIRY_CNT_1M_CR22,ENQUIRY_CNT_3M_CR22,ENQUIRY_CNT_6M_CR22,ENQUIRY_CNT_12M_CR22,ENQUIRY_CNT_CR21,ENQUIRY_CNT_1M_CR21,ENQUIRY_CNT_3M_CR21,ENQUIRY_CNT_6M_CR21,ENQUIRY_CNT_12M_CR21,ENQUIRY_MAX_AMT_1M,ENQUIRY_MAX_AMT_3M,ENQUIRY_MAX_AMT_6M,ENQUIRY_MAX_AMT_12M,OPEN_ACCT_CR22,CLOSED_ACCT_CR22,MORTGAGE_CR22,CREDIT_CARD_CR22,TELCO_CR22,UTILITIES_CR22,OVERDRAFT_CR22,PERSONAL_LOAN_CR22,AUTO_LOAN_CR22,CREDIT_CARD_AMT_CR22,OVERDRAFT_AMT_CR22,PERSONAL_LOAN_AMT_CR22,AUTO_LOAN_AMT_CR22,OPEN_ACCT_CR21,CLOSED_ACCT_CR21,MORTGAGE_CR21,CREDIT_CARD_CR21,TELCO_CR21,UTILITIES_CR21,OVERDRAFT_CR21,PERSONAL_LOAN_CR21,AUTO_LOAN_CR21,CREDIT_CARD_AMT_CR21,OVERDRAFT_AMT_CR21,PERSONAL_LOAN_AMT_CR21,AUTO_LOAN_AMT_CR21,DEFAULT_CNT_CR22,DEFAULT_OPEN_CNT_CR22,DEFAULT_CNT_3M_CR22,DEFAULT_CNT_6M_CR22,DEFAULT_CNT_12M_CR22,DEFAULT_UNPAID_AMT_CR22,DEFAULT_CNT_CR21,DEFAULT_OPEN_CNT_CR21,DEFAULT_UNPAID_AMT_CR21,DEFAULT_CNT_3M_CR21,DEFAULT_CNT_6M_CR21,DEFAULT_CNT_12M_CR21,AGENCY_AMT,MTH_TILL_AGENCY
0,2975536526,queries 12 Months 1-2,2,2,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,NaN,NaN
1,2999806086,queries 12 Months 1-2,2,2,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,NaN,NaN
2,2800835182,queries 12 Months 4-5,5,5,3.00,2.00,2.00,3.00,3.00,6.00,1.00,1.00,2.00,2.00,0.00,0.00,0.00,0.00,7.00,2.00,6.00,1.00,0.00,0.00,0.00,0.00,0.00,17500.00,0.00,0.00,0.00,8.00,1.00,6.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,NaN,NaN
3,2975640662,queries 12 Months 4-5,2,4,1.00,1.00,1.00,1.00,1.00,4.00,1.00,1.00,1.00,3.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,1.00,459.00,0.00,0.00,0.00,NaN,NaN
4,2979232086,queries 12 Months 1-2,2,2,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,60.01,16.00


In [212]:
df_bureau.shape

(91849, 58)

In [213]:
df_bureau.columns

Index(['Customer', 'BUREAU_ENQUIRIES_12_MONTHS', 'ENQUIRIES_6M',
       'ENQUIRIES_12M', 'ENQUIRY_CNT_CR22', 'ENQUIRY_CNT_1M_CR22',
       'ENQUIRY_CNT_3M_CR22', 'ENQUIRY_CNT_6M_CR22', 'ENQUIRY_CNT_12M_CR22',
       'ENQUIRY_CNT_CR21', 'ENQUIRY_CNT_1M_CR21', 'ENQUIRY_CNT_3M_CR21',
       'ENQUIRY_CNT_6M_CR21', 'ENQUIRY_CNT_12M_CR21', 'ENQUIRY_MAX_AMT_1M',
       'ENQUIRY_MAX_AMT_3M', 'ENQUIRY_MAX_AMT_6M', 'ENQUIRY_MAX_AMT_12M',
       'OPEN_ACCT_CR22', 'CLOSED_ACCT_CR22', 'MORTGAGE_CR22',
       'CREDIT_CARD_CR22', 'TELCO_CR22', 'UTILITIES_CR22', 'OVERDRAFT_CR22',
       'PERSONAL_LOAN_CR22', 'AUTO_LOAN_CR22', 'CREDIT_CARD_AMT_CR22',
       'OVERDRAFT_AMT_CR22', 'PERSONAL_LOAN_AMT_CR22', 'AUTO_LOAN_AMT_CR22',
       'OPEN_ACCT_CR21', 'CLOSED_ACCT_CR21', 'MORTGAGE_CR21',
       'CREDIT_CARD_CR21', 'TELCO_CR21', 'UTILITIES_CR21', 'OVERDRAFT_CR21',
       'PERSONAL_LOAN_CR21', 'AUTO_LOAN_CR21', 'CREDIT_CARD_AMT_CR21',
       'OVERDRAFT_AMT_CR21', 'PERSONAL_LOAN_AMT_CR21', 'AUTO_LOAN_AMT_C

In [214]:
df_bureau.dtypes

Customer                        int64
BUREAU_ENQUIRIES_12_MONTHS     object
ENQUIRIES_6M                    int64
ENQUIRIES_12M                   int64
ENQUIRY_CNT_CR22              float64
ENQUIRY_CNT_1M_CR22           float64
ENQUIRY_CNT_3M_CR22           float64
ENQUIRY_CNT_6M_CR22           float64
ENQUIRY_CNT_12M_CR22          float64
ENQUIRY_CNT_CR21              float64
ENQUIRY_CNT_1M_CR21           float64
ENQUIRY_CNT_3M_CR21           float64
ENQUIRY_CNT_6M_CR21           float64
ENQUIRY_CNT_12M_CR21          float64
ENQUIRY_MAX_AMT_1M            float64
ENQUIRY_MAX_AMT_3M            float64
ENQUIRY_MAX_AMT_6M            float64
ENQUIRY_MAX_AMT_12M           float64
OPEN_ACCT_CR22                float64
CLOSED_ACCT_CR22              float64
MORTGAGE_CR22                 float64
CREDIT_CARD_CR22              float64
TELCO_CR22                    float64
UTILITIES_CR22                float64
OVERDRAFT_CR22                float64
PERSONAL_LOAN_CR22            float64
AUTO_LOAN_CR

In [215]:
df_bureau.isna().sum()

Customer                          0
BUREAU_ENQUIRIES_12_MONTHS        0
ENQUIRIES_6M                      0
ENQUIRIES_12M                     0
ENQUIRY_CNT_CR22               7425
ENQUIRY_CNT_1M_CR22            7425
ENQUIRY_CNT_3M_CR22            7425
ENQUIRY_CNT_6M_CR22            7425
ENQUIRY_CNT_12M_CR22           7425
ENQUIRY_CNT_CR21               7425
ENQUIRY_CNT_1M_CR21            7425
ENQUIRY_CNT_3M_CR21            7425
ENQUIRY_CNT_6M_CR21            7425
ENQUIRY_CNT_12M_CR21           7425
ENQUIRY_MAX_AMT_1M             7425
ENQUIRY_MAX_AMT_3M             7425
ENQUIRY_MAX_AMT_6M             7425
ENQUIRY_MAX_AMT_12M            7425
OPEN_ACCT_CR22                 7425
CLOSED_ACCT_CR22               7425
MORTGAGE_CR22                  7425
CREDIT_CARD_CR22               7425
TELCO_CR22                     7425
UTILITIES_CR22                 7425
OVERDRAFT_CR22                 7425
PERSONAL_LOAN_CR22             7425
AUTO_LOAN_CR22                 7425
CREDIT_CARD_AMT_CR22        

In [216]:
df_bureau.duplicated().sum()

np.int64(0)

In [217]:
cols_to_drop = [
    "AGENCY_AMT",
    "MTH_TILL_AGENCY"
]

df_bureau = df_bureau.drop(columns=cols_to_drop)
df_bureau.head()

,Customer,BUREAU_ENQUIRIES_12_MONTHS,ENQUIRIES_6M,ENQUIRIES_12M,ENQUIRY_CNT_CR22,ENQUIRY_CNT_1M_CR22,ENQUIRY_CNT_3M_CR22,ENQUIRY_CNT_6M_CR22,ENQUIRY_CNT_12M_CR22,ENQUIRY_CNT_CR21,ENQUIRY_CNT_1M_CR21,ENQUIRY_CNT_3M_CR21,ENQUIRY_CNT_6M_CR21,ENQUIRY_CNT_12M_CR21,ENQUIRY_MAX_AMT_1M,ENQUIRY_MAX_AMT_3M,ENQUIRY_MAX_AMT_6M,ENQUIRY_MAX_AMT_12M,OPEN_ACCT_CR22,CLOSED_ACCT_CR22,MORTGAGE_CR22,CREDIT_CARD_CR22,TELCO_CR22,UTILITIES_CR22,OVERDRAFT_CR22,PERSONAL_LOAN_CR22,AUTO_LOAN_CR22,CREDIT_CARD_AMT_CR22,OVERDRAFT_AMT_CR22,PERSONAL_LOAN_AMT_CR22,AUTO_LOAN_AMT_CR22,OPEN_ACCT_CR21,CLOSED_ACCT_CR21,MORTGAGE_CR21,CREDIT_CARD_CR21,TELCO_CR21,UTILITIES_CR21,OVERDRAFT_CR21,PERSONAL_LOAN_CR21,AUTO_LOAN_CR21,CREDIT_CARD_AMT_CR21,OVERDRAFT_AMT_CR21,PERSONAL_LOAN_AMT_CR21,AUTO_LOAN_AMT_CR21,DEFAULT_CNT_CR22,DEFAULT_OPEN_CNT_CR22,DEFAULT_CNT_3M_CR22,DEFAULT_CNT_6M_CR22,DEFAULT_CNT_12M_CR22,DEFAULT_UNPAID_AMT_CR22,DEFAULT_CNT_CR21,DEFAULT_OPEN_CNT_CR21,DEFAULT_UNPAID_AMT_CR21,DEFAULT_CNT_3M_CR21,DEFAULT_CNT_6M_CR21,DEFAULT_CNT_12M_CR21
0,2975536526,queries 12 Months 1-2,2,2,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
1,2999806086,queries 12 Months 1-2,2,2,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
2,2800835182,queries 12 Months 4-5,5,5,3.00,2.00,2.00,3.00,3.00,6.00,1.00,1.00,2.00,2.00,0.00,0.00,0.00,0.00,7.00,2.00,6.00,1.00,0.00,0.00,0.00,0.00,0.00,17500.00,0.00,0.00,0.00,8.00,1.00,6.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
3,2975640662,queries 12 Months 4-5,2,4,1.00,1.00,1.00,1.00,1.00,4.00,1.00,1.00,1.00,3.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,1.00,459.00,0.00,0.00,0.00
4,2979232086,queries 12 Months 1-2,2,2,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00


In [218]:
pd.set_option("display.max_rows", None)
pd.set_option("display.max_columns", None)


In [219]:
for col in df_bureau.select_dtypes(include='object').columns:
    print(col, "-->", df_bureau[col].unique())

BUREAU_ENQUIRIES_12_MONTHS --> ['queries 12 Months 1-2' 'queries 12 Months 4-5' 'queries 12 Months 3'
 'queries 12 Months 6' 'queries 12 Months 8-11' 'queries 12 Months 12-14'
 'queries 12 Months 7' 'queries 12 Months 14+']


In [220]:
bureau_enq_12m_map = {
    'queries 12 Months 1-2': '1-2',
    'queries 12 Months 3': '3',
    'queries 12 Months 4-5': '4-5',
    'queries 12 Months 6': '6-7',
    'queries 12 Months 7': '6-7',
    'queries 12 Months 8-11': '8-11',
    'queries 12 Months 12-14': '12+',
    'queries 12 Months 14+': '14+'
}

df_bureau['BUREAU_ENQUIRIES_12_MONTHS'] = (
    df_bureau['BUREAU_ENQUIRIES_12_MONTHS']
    .map(bureau_enq_12m_map)
    .fillna('Missing')
)


In [221]:
for col in df_bureau.select_dtypes(include='object').columns:
    print(col, "-->", df_bureau[col].unique())

BUREAU_ENQUIRIES_12_MONTHS --> ['1-2' '4-5' '3' '6-7' '8-11' '12+' '14+']


-----------------------------------------------------------------------------------------------------------

<div style="text-align:center;">
  <span style="color:Green; font-size:30px; font-weight:bold;">
    Merging Customer , Loan , Bureau Table
  </span>
</div>

In [222]:
crm_1 = pd.merge(df_customer, df_loan, how="left", on=["Customer"])
crm_1.shape

(91849, 34)

In [223]:
crm_1.head()

,OUTCOME,Customer,LICENCE_AVAILABLE,RESIDENTIAL,CD_OCCUPATION,CD_STS_RESIDENTIAL,HAD_SERVICES,DOC_TYPE,VISA_TYPE,EMPLOYED_STATUS,EMPLOYED_TYPE,EMPLOYED_NO_MTHS,NO_ADDRESS_AT_MONTHS,APPLICANT_AGE,APPLICANT_AGE_AVG,RISK_SCORE,BUREAU_DEFAULT,DEFAULT_AMT,DEFAULTS_12M,DEROGATORIES,NO_SCORE,SCORECARD,SCORE_CR22,SCORE_BAND_CR22,SCORECARD_CR21,RISK_RELATIVE_CR21,RISK_RELATIVE_ODDS_CR21,Late_Payment_30DPD_Last_12M,Credit_Card_Payment_Failure_Count,Personal_Loan_Payment_Failure_Count,Recent_Payment_Irregularity_Flag,Late_Payment_30DPD_Last_24M,Cumulative_Delinquency_Exposure,Long_Term_Payment_Delinquency_Count
0,Good,2975536526,Missing,Rented,employed,R,1.00,Intl Passport and Visa,Work Visa,employed,Missing,1.00,1.00,30-34,32,628,Missing,0,0,0,724,TAR1A,586.00,High,CTS1A,9.14,37.79,0,-1,-1,0,0,0,0
1,Good,2999806086,yes,Rented,employed,R,1.00,AU Driver Licence,Resident,employed,Missing,86.00,49.00,30-34,32,813,Missing,0,0,0,775,TAR1A,636.00,Medium,CTS1A,13.01,53.82,0,1,-1,24,0,4000,23
2,Good,2800835182,yes,Owned,employed,O,1.00,AU Driver Licence,Resident,employed,Missing,61.00,129.00,35-44,40,1012,Missing,0,0,0,815,SFJR1A,952.00,Low,CTSSH,17.17,71.01,0,0,-1,24,0,380277,24
3,Good,2975640662,Missing,Rented,employed,R,1.00,Intl Passport and Visa,Resident,employed,Missing,15.00,60.00,25 - 29,27,415,1-1000,459,0,1,328,HSHSOL,534.00,High,CTSAD,-1.70,2.43,0,-1,-1,0,0,0,0
4,Good,2979232086,Missing,Living_With_Family,unemployed,B,1.00,Intl Passport and Visa,Work Visa,unemployed,unemployed,0.00,1.00,18-24,21,556,Missing,0,0,0,692,TAR1A,419.00,High,CTS1A,7.32,30.27,0,-1,-1,0,0,0,0


In [224]:
crm = pd.merge(crm_1, df_bureau, how="left", on=["Customer"])
crm.shape

(91849, 89)

In [225]:
crm.columns

Index(['OUTCOME', 'Customer', 'LICENCE_AVAILABLE', 'RESIDENTIAL',
       'CD_OCCUPATION', 'CD_STS_RESIDENTIAL', 'HAD_SERVICES', 'DOC_TYPE',
       'VISA_TYPE', 'EMPLOYED_STATUS', 'EMPLOYED_TYPE', 'EMPLOYED_NO_MTHS',
       'NO_ADDRESS_AT_MONTHS', 'APPLICANT_AGE', 'APPLICANT_AGE_AVG',
       'RISK_SCORE', 'BUREAU_DEFAULT', 'DEFAULT_AMT', 'DEFAULTS_12M',
       'DEROGATORIES', 'NO_SCORE', 'SCORECARD', 'SCORE_CR22',
       'SCORE_BAND_CR22', 'SCORECARD_CR21', 'RISK_RELATIVE_CR21',
       'RISK_RELATIVE_ODDS_CR21', 'Late_Payment_30DPD_Last_12M',
       'Credit_Card_Payment_Failure_Count',
       'Personal_Loan_Payment_Failure_Count',
       'Recent_Payment_Irregularity_Flag', 'Late_Payment_30DPD_Last_24M',
       'Cumulative_Delinquency_Exposure',
       'Long_Term_Payment_Delinquency_Count', 'BUREAU_ENQUIRIES_12_MONTHS',
       'ENQUIRIES_6M', 'ENQUIRIES_12M', 'ENQUIRY_CNT_CR22',
       'ENQUIRY_CNT_1M_CR22', 'ENQUIRY_CNT_3M_CR22', 'ENQUIRY_CNT_6M_CR22',
       'ENQUIRY_CNT_12M_CR22', 'EN

--------------------------------------------------------------------------------------------------------------------------------

<div style="text-align:center;">
  <span style="color:Green; font-size:30px; font-weight:bold;">
    credit_risk_master_table
  </span>
</div>

In [226]:
crm.to_csv("credit_risk_master_table_2.csv", index=False)


**EDA and model training were done in separate notebooks to maintain a clear and structured workflow.**

-----------------------------------------------------------